v2/cmd/asoctl/internal/importing/resource_importer.go

```diff
@@ -160,94 +121,220 @@ func (ri *ResourceImporter) Import(
 		resources = append(resources, importer.Resource())
 	}
 
-	ri.report.WriteToLog(ri.log)
-
 	return &ResourceImportResult{
 		resources: resources,
-	}, kerrors.NewAggregate(errs)
+	}, nil
 }
 
-func (ri *ResourceImporter) ImportResource(ctx context.Context, rsrc ImportableResource) error {
-	// Import it
-	gk := rsrc.GroupKind()
-	name := fmt.Sprintf("%s %s", gk, rsrc.Name())
-	bar := ri.progress.AddBar(
-		0, // zero total because we don't know how much work will be needed
-		mpb.BarRemoveOnComplete(),
-		mpb.PrependDecorators(
-			decor.Name(name, decor.WCSyncSpaceR)),
-		mpb.AppendDecorators(decor.Percentage(decor.WC{W: 5})))
+// queueUniqueImporters reads from the candicates channel, putting each importer onto pending exactly once.
+// This ensures each distinct resource is only imported once, regardless of how many times we
+// encounter it. We also proactively buffer the pending channel to avoid blocking. Any fixed size channel
+// would risk deadlock for a sufficiently large resource graph.
+func (ri *ResourceImporter) queueUniqueImporters(
+	candidates chan ImportableResource,
+	pending chan ImportableResource,
+	progress chan progressDelta,
+) {
+	seen := set.Make[string]()
+	var queue []ImportableResource
+	var current ImportableResource = nil
+	for {
+		// Dequeue from our internal buffer if needed
+		if current == nil && len(queue) > 0 {
+			current = queue[0]
+			queue = queue[1:]
+		}
 
-	defer func() {
-		// Ensure the progress bar is always updated when we're done
-		// We're done with this resource, so remove it from the progress bar
-		bar.Increment()
-		bar.SetTotal(bar.Current(), true)
-	}()
+		// If we have a current importable to send, use the pending queue
+		var upstream chan ImportableResource = nil
+		if current != nil {
+			upstream = pending
+		}
 
-	pending, err := rsrc.Import(ctx, bar)
-	if err != nil {
-		var skipped *ImportSkippedError
-		if errors.As(err, &skipped) {
-			ri.log.V(1).Info(
-				"Skipped",
-				"kind", gk,
-				"name", rsrc.Name(),
-				"because", skipped.Because)
-			ri.report.AddSkippedImport(rsrc, skipped.Because)
-			return nil
+		select {
+		case rsrc := <-candidates:
+			if seen.Contains(rsrc.Id()) {
+				// We've already seen this resource (we've already queued it for import)
+				// So remove it from our count of work to be done
+				ri.log.V(2).Info("Skipping duplicate import", "resource", rsrc.Id())
+				progress <- progressDelta{total: -1}
+			} else {
+				// Remember we've seen this, and add it to our queue
+				seen.Add(rsrc.Id())
+				queue = append(queue, rsrc)
+				ri.log.V(2).Info("Buffering import", "resource", rsrc.Id())
+			}
+		case upstream <- current:
+			// We've sent the current importable, so clear it
+			ri.log.V(2).Info("Queued import", "resource", current.Id())
+			current = nil
 		}
+	}
+}
 
-		ri.log.Error(err,
-			"Failed",
-			"kind", gk,
-			"name", rsrc.Name())
+func (ri *ResourceImporter) importWorker(
+	ctx context.Context,
+	pending chan ImportableResource,
```

Comment:
Do these channels need to be bi-directional?

Reply:
Yeah, they probably should be. Will change.